alternator: hide internal tags from users

nyh opened the pull request:
The "tags" mechanism in Alternator is a convenient way to attach metadata to Alternator tables. Recently we have started using it more and more for internal metadata storage:

  * UpdateTimeToLive stores the attribute in a tag system:ttl_attribute
  * CreateTable stores provisioned throughput in tags system:provisioned_rcu and system:provisioned_wcu
  * CreateTable stores the table's creation time in a tag called system:table_creation_time.

We do not want any of these internal tags to be visible to a ListTagsOfResource request, because if they are visible (as before this patch), systems such as Terraform can get confused when they suddenly see a tag which they didn't set - and may even attempt to delete it (as reported in issue #24098).

Moreover, we don't want any of these internal tags to be writable with TagResource or UntagResource: If a user wants to change the TTL setting they should do it via UpdateTimeToLive - not by writing directly to tags.

So in this patch we forbid read or write to *any* tag that begins with the "system:" prefix, except one: "system:write_isolation". That tag is deliberately intended to be writable by the user, as a configuration mechanism, and is never created internally by Scylla. We should have perhaps chosen a different prefix for configurable vs. internal tags, or chosen more unique prefixes - but let's not change these historic names now.

This patch also adds regression tests for the internal tags features, failing before this patch and passing after:
1. internal tags, specifically system:ttl_attribute, are not visible in ListTagsOfResource, and cannot be modified by TagResource or UntagResource.
2. system:write_isolation is not internal, and be written by either TagResource or UntagResource, and read with ListTagsOfResource.

Fixes #24098.

This issue was reported by (internal) users, and the fix is very simple, so it's probably worth backporting.

nyh commented on alternator/executor.cc:
```diff
@@ -1006,6 +1006,17 @@ void rmw_operation::set_default_write_isolation(std::string_view value) {
+// are deliberately intended to be set and read by the user, so are not
+// considered "internal".
+static bool tag_key_is_internal(std::string_view tag_key) {
+    return tag_key.starts_with("system:") &&
```
If I were to do that, I would also want the constants like WRITE_ISOLATION_TAG_KEY et al. to use that new tag prefix string - right now they have the "system:..." in them too. I don't want to change too much existing code.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17436/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17436/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17436/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:small_blue_diamond: alternator/test_ttl
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5329/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16580/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8497/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17436/)
#### Build Details:
- Duration: 9 hr 6 min
- Builder: i-0817242b4b552d11c (m5d.16xlarge)

ScyllaPiotr commented on test/alternator/test_ttl.py:
```diff
@@ -809,3 +808,42 @@ def test_ttl_expiration_long(dynamodb, waits_for_expiration):
+            # Verify that TTL is set for this table, but no extra tags
+            # like "system:ttl_attribute" (or anything else) are visible
+            # in ListTagsOfResource:
+            assert client.describe_time_to_live(TableName=table.name)['TimeToLiveDescription'] == {'TimeToLiveStatus': 'ENABLED', 'AttributeName': 'x'}
```
According to [boto3 doc](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb/client/update_time_to_live.html):
> It can take up to one hour for the change to fully process.

However, my few runs of the test against AWS were all successful, so I guess this is safe.

nyh replied:
I didn't notice that, but now that I look at it, I'm not exactly sure what this statement means. Not just this tests, but many other tests in test_ttl.py use update_time_to_live and then immediately describe_time_to_live, and all of them worked on DynamoDB without a problem. I am guessing that describe_time_to_live() is will always (de facto, at least) return the up-to-date information, but what may take up to an hour is for the background expiration process to actually start working.

ScyllaPiotr replied:
I have exactly the same impression.

ScyllaPiotr commented on test/alternator/test_ttl.py:
```diff
@@ -809,3 +808,42 @@ def test_ttl_expiration_long(dynamodb, waits_for_expiration):
+    client = test_table.meta.client
+    arn = client.describe_table(TableName=test_table.name)['Table']['TableArn']
+    with pytest.raises(ClientError, match='ValidationException.*internal'):
+        client.tag_resource(ResourceArn=arn, Tags=[{'Key': tag_name, 'Value': 'x'}])
```
It would make sense to attempt to write a tag value that is valid. Is `x` valid?

nyh commented on test/alternator/test_tag.py:
```diff
@@ -144,6 +144,13 @@ def test_tag_resource_write_isolation_values(scylla_only, test_table):
+    assert 'system:write_isolation' in keys
+    # Finally remove the system:write_isolation tag so to not modify the
+    # default behavior of test_table.
+    test_table.meta.client.untag_resource(ResourceArn=arn, TagKeys=['system:write_isolation'])
```
I actually is mentioned in the commit message, which isn't short ;-) I said there that 
"system:write_isolation is not internal, and [can] be written by either TagResource or UntagResource, and read with ListTagsOfResource."
(the word "can" was missing).

nyh replied:
Oh, you meant just the "remove the tag"? Yes, that was a bug in the test! I'll mention it now.

nyh commented:
Rebased and updated commit message (didn't change the cover letter, which will get ignored anyway) to add a paragraph about fixing a pre-existing bug in the test that we're modifying (thanks to @ScyllaPiotr for noticing I forgot to mention it).

nyh commented on test/alternator/test_ttl.py:
```diff
@@ -809,3 +808,42 @@ def test_ttl_expiration_long(dynamodb, waits_for_expiration):
+        AttributeDefinitions=[{ 'AttributeName': 'p', 'AttributeType': 'S' }]
+        ) as table:
+            client = table.meta.client
+            client.update_time_to_live(TableName=table.name,
```
As they said in the Perl world, "there's more than one way to do it" (https://en.wiktionary.org/wiki/TMTOWTDI) :-)

If this was the only test that calls `describe_time_to_live` without any retries, I would think you are right. But we already have many other tests like this, and none of them failed on DynamoDB, so I think it's fine to check like this - and assume that the phrase of "it can take up to one hour" means the time until the old TTL background process stops and the new one starts, and _not_ the time until `describe_time_to_live` can return the information. In practice, that seems immediate.

nyh replied:
Then what's the point of checking the result of a statement that always returns its parameter?
Maybe I shouldn't just shouldn't test this aspect at all? Just assume that after I called update_time_to_live, it worked. And that's it. The extra call at least checked something extra, some "sanity" that we reached the intended state. If it's the same call I think we don't need to check at all.

I agree that in theory the documentation may suggest that describe_time_to_live can take time to realize what changed, but the tests demonstrate that this is not the case in DynamoDB, and it's important that we know that (but unrelated to the tag hiding thing, of course, I mean in the other TTL tests).

ScyllaPiotr replied:
> Then what's the point of checking the result of a statement that always returns its parameter? Maybe I shouldn't just shouldn't test this aspect at all? Just assume that after I called update_time_to_live, it worked. And that's it. The extra call at least checked something extra, some "sanity" that we reached the intended state. If it's the same call I think we don't need to check at all.

I'm not sure I understand what you mean.
The doc. says that _A successful UpdateTimeToLive call returns the current TimeToLiveSpecification_. If it happens to be exactly the same as we requested it to be, we're good and all the sanity checks on the DBE side were green.

> I agree that in theory the documentation may suggest that describe_time_to_live can take time to realize what changed, but the tests demonstrate that this is not the case in DynamoDB, and it's important that we know that (but unrelated to the tag hiding thing, of course, I mean in the other TTL tests).

Agreed.

nyh commented:
@scylladb/scylla-maint please consider merging, this is a prerequisite for https://github.com/scylladb/scylladb/pull/24007.

We have passing tests and two approvals (I have a slight disagreement with @ScyllaPiotr about some sanity check that I could have done in a different way, but he agreed it's not blocking, and my way uses a pattern that is already heavily used in that file so we can later consider changing all the uses and not just the new one).
